File: backend/internal/adapter/delivery/ws/dnp/codec.go
package dnp

import (
	"encoding/binary"
	"errors"

	"github.com/flynn/noise"
)

// MaxFrameLen — верхняя граница payload одного кадра (совпадает с ws.maxMessageSize).
const MaxFrameLen = 1 << 20

// Формат кадра: u32 big-endian длина payload + payload. Над WS префикс избыточен
// (границы даёт сам WS), но закладываем его единообразно ради будущего сырого-TCP
// носителя (см. спеку DNP §4/§10.8). Хендшейк: payload = сырое Noise-сообщение;
// транспорт: payload = зашифрованный кадр.
func FrameLen(payload []byte) []byte {
	out := make([]byte, 4+len(payload))
	binary.BigEndian.PutUint32(out[:4], uint32(len(payload)))
	copy(out[4:], payload)
	return out
}

// UnframeLen снимает длину и валидирует, что она ровно соответствует остатку.
func UnframeLen(raw []byte) ([]byte, error) {
	if len(raw) < 4 {
		return nil, errors.New("dnp: short frame header")
	}
	n := binary.BigEndian.Uint32(raw[:4])
	if n > MaxFrameLen {
		return nil, errors.New("dnp: frame too large")
	}
	if int(n) != len(raw)-4 {
		return nil, errors.New("dnp: frame length mismatch")
	}
	return raw[4:], nil
}

// Seal/Open — только крипта (без длины), AD пустой. Для случаев, где длина уже снята
// (напр. auth-кадр после readWSFramed).
func Seal(cs *noise.CipherState, plaintext []byte) ([]byte, error) {
	return cs.Encrypt(nil, nil, plaintext)
}
func Open(cs *noise.CipherState, ciphertext []byte) ([]byte, error) {
	return cs.Decrypt(nil, nil, ciphertext)
}

// EncryptFrame шифрует прикладной кадр и оборачивает длиной: [len][Encrypt(pt)].
func EncryptFrame(cs *noise.CipherState, plaintext []byte) ([]byte, error) {
	ct, err := Seal(cs, plaintext)
	if err != nil {
		return nil, err
	}
	return FrameLen(ct), nil
}

// DecryptFrame снимает длину и расшифровывает: Open(unframe(raw)).
func DecryptFrame(cs *noise.CipherState, raw []byte) ([]byte, error) {
	ct, err := UnframeLen(raw)
	if err != nil {
		return nil, err
	}
	return Open(cs, ct)
}

File: web-client/src/core/net/dnp/codec.ts
import type { CipherState } from './noise/symmetricState'

// Верхняя граница payload одного кадра (совпадает с бэкендом ws maxMessageSize).
export const MAX_FRAME_LEN = 1 << 20
const EMPTY_AD = new Uint8Array(0)

// Формат кадра: u32 big-endian длина + payload. Над WS префикс избыточен (границы
// даёт сам WS), но закладываем единообразно с бэком ради будущего сырого-TCP носителя.
export function frameLen(payload: Uint8Array): Uint8Array {
  const out = new Uint8Array(4 + payload.length)
  new DataView(out.buffer).setUint32(0, payload.length, false) // big-endian
  out.set(payload, 4)
  return out
}

export function unframeLen(raw: Uint8Array): Uint8Array {
  if (raw.length < 4) throw new Error('dnp: short frame header')
  const n = new DataView(raw.buffer, raw.byteOffset, 4).getUint32(0, false)
  if (n > MAX_FRAME_LEN) throw new Error('dnp: frame too large')
  if (n !== raw.length - 4) throw new Error('dnp: frame length mismatch')
  return raw.subarray(4)
}

// Транспортный кадр: [len][encrypt(pt)], AD пустой.
export function sealFrame(cs: CipherState, plaintext: Uint8Array): Uint8Array {
  return frameLen(cs.encryptWithAd(EMPTY_AD, plaintext))
}

export function openFrame(cs: CipherState, raw: Uint8Array): Uint8Array {
  return cs.decryptWithAd(EMPTY_AD, unframeLen(raw))
}
